src/sync/mu.cr

```diff
@@ -0,0 +1,380 @@
+        return if success
+      end
+
+      # must try to wakeup a waiter
+      runlock_slow
+    end
+
+    def unlock_slow : Nil
+      unlock_slow_impl(sub_on_release: WLOCK)
+    end
+
+    def runlock_slow : Nil
+      unlock_slow_impl(sub_on_release: RLOCK)
+    end
+
+    private def unlock_slow_impl(sub_on_release) : Nil
+      attempts = 0
+
+      while true
+        word = @word.get(:relaxed)
+
+        if (word & WAITING) == 0 || (word & DESIG_WAKER) != 0 || (word & RMASK) > RLOCK
+          # no waiters, there is a designated waker (no need to wake another
+          # one), or there are still readers, try release lock
+          word, success = @word.compare_and_set(word, word - sub_on_release, :release, :relaxed)
+          return if success
+        elsif (word & SPINLOCK) == 0
+          # there might be a waiter, and no designated waker, try to acquire
+          # spinlock, and release the lock (early)
+          _, success = @word.compare_and_set(word, (word | SPINLOCK | DESIG_WAKER) &- sub_on_release, :acquire_release, :relaxed)
+          if success
+            # spinlock is held, resume a single writer, or resume all readers
+            wake = Crystal::PointerLinkedList(Waiter).new
+            writer_waiting = 0_u32
+
+            if first_waiter = @waiters.shift?
+              wake.push(first_waiter)
+
+              if first_waiter.value.reader?
+                @waiters.each do |waiter|
+                  if waiter.value.reader?
+                    @waiters.delete(waiter)
+                    wake.push(waiter)
+                  else
+                    # found a writer, prevent new readers from locking
+                    writer_waiting = WRITER_WAITING
+                  end
+                end
+              end
+            end
+
+            # update flags
+            clear = 0_u32
+            clear |= DESIG_WAKER if wake.empty? # nothing to wake => no designated waker
+            clear |= WAITING if @waiters.empty? # no more waiters => nothing waiting
+
+            release_spinlock(set: writer_waiting, clear: clear)
+
+            wake.consume_each do |waiter|
+              waiter.value.wake
+            end
+
+            return
+          end
+        end
+
+        attempts = Thread.delay(attempts)
+      end
+    end
+
+    def held? : Bool
+      word = @word.get(:relaxed)
+      (word & WLOCK) != 0
+    end
+
+    def rheld? : Bool
+      word = @word.get(:relaxed)
+      (word & RMASK) != 0
+    end
+
+    private def release_spinlock(set = 0_u32, clear = 0_u32)
+      word = @word.get(:relaxed)
+
+      while true
+        word, success = @word.compare_and_set(word, (word | set) & ~(SPINLOCK | clear), :release, :relaxed)
+        return if success
+      end
+    end
+
+    protected def try_transfer(wake : Pointer(Crystal::PointerLinkedList(Waiter)), first_waiter : Pointer(Waiter), all_readers : Bool) : Nil
```

Comment:
The `#try_transfer` and `#transfer` shall actually come with the `Sync::ConditionVariable` PR.